Add GLFW_CENTER_CURSOR window hint

liam-middlebrook opened the pull request:
Implement a hint for positioning the cursor in the center of the window.

Fixes #749

shurcooL commented:
Does this flag achieve something that can't be done simply by using existing APIs to get the monitor size, calculate its center, and position the cursor?

elmindreda commented on src/cocoa_window.m:
```diff
@@ -236,7 +236,8 @@ - (void)windowDidResize:(NSNotification *)notification
     if (_glfw.ns.disabledCursorWindow == window)
-        centerCursor(window);
+	if (_glfw.hints.window.centerCursor)
+            centerCursor(window);
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/cocoa_window.m:
```diff
@@ -251,7 +252,8 @@ - (void)windowDidMove:(NSNotification *)notification
     if (_glfw.ns.disabledCursorWindow == window)
-        centerCursor(window);
+	if (_glfw.hints.window.centerCursor)
+            centerCursor(window);
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/cocoa_window.m:
```diff
@@ -277,7 +279,8 @@ - (void)windowDidDeminiaturize:(NSNotification *)notification
     if (_glfw.ns.disabledCursorWindow == window)
-        centerCursor(window);
+	if (_glfw.hints.window.centerCursor)
+            centerCursor(window);
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/cocoa_window.m:
```diff
@@ -1453,7 +1457,8 @@ void _glfwPlatformSetCursorMode(_GLFWwindow* window, int mode)
                                   &_glfw.ns.restoreCursorPosY);
-        centerCursor(window);
+	if (_glfw.hints.window.centerCursor)
+            centerCursor(window);
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/win32_window.c:
```diff
@@ -1472,7 +1475,10 @@ void _glfwPlatformSetCursorMode(_GLFWwindow* window, int mode)
+	if (_glfw.hints.window.centerCursor)
+	{
+            centerCursor(window);
+	}
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/x11_window.c:
```diff
@@ -2063,7 +2066,10 @@ void _glfwPlatformPollEvents(void)
+	if(_glfw.hints.window.centerCursor)
+	{
+            centerCursor(_glfw.x11.disabledCursorWindow);
+	}
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/x11_window.c:
```diff
@@ -2137,7 +2143,10 @@ void _glfwPlatformSetCursorMode(_GLFWwindow* window, int mode)
+	if(_glfw.hints.window.centerCursor)
+	{
+            centerCursor(window);
+	}
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/win32_window.c:
```diff
@@ -1017,7 +1017,10 @@ int _glfwPlatformCreateWindow(_GLFWwindow* window,
+	if (_glfw.hints.window.centerCursor)
+	{
+            centerCursor(window);
+	}
```
This should be routed via the `_GLFWwndconfig` argument.

elmindreda commented on src/cocoa_window.m:
```diff
@@ -1053,7 +1056,8 @@ int _glfwPlatformCreateWindow(_GLFWwindow* window,
 
-        centerCursor(window);
+	if (_glfw.hints.window.centerCursor)
+            centerCursor(window);
```
This should be routed via the `_GLFWwndconfig` argument.

elmindreda commented on src/x11_window.c:
```diff
@@ -1598,7 +1598,10 @@ int _glfwPlatformCreateWindow(_GLFWwindow* window,
+	if(_glfw.hints.window.centerCursor)
+	{
+            centerCursor(window);
+	}
```
This should be routed via the `_GLFWwndconfig` argument.

elmindreda commented on docs/window.dox:
```diff
@@ -892,6 +893,9 @@ the user or with @ref glfwIconifyWindow.
 the user or with @ref glfwMaximizeWindow.
 
+`GLFW_CENTER_CURSOR` indicates whether the cursor should be moved to the center
+of the window during creation.
```
`of the full screen window during creation.  This is ignored for windowed mode windows.` or something like that.

elmindreda commented on README.md:
```diff
@@ -105,6 +105,7 @@ information on what to include when reporting a bug.
 - Bugfix: `GLFW_INCLUDE_VULKAN` could not be combined with the corresponding
           OpenGL and OpenGL ES header macros
 - [Win32] Bugfix: Undecorated windows could not be iconified by the user (#861)
+- Added 'GLFW_CENTER_CURSOR' window hint
```
`for controlling cursor centering` or something like that.

liam-middlebrook commented:
I rebased around a handful of changes. Let me know if there's anything else you need. I believe I've removed anything that related to disabled cursors.

elmindreda commented on src/x11_window.c:
```diff
@@ -2160,7 +2163,6 @@ void _glfwPlatformPollEvents(void)
     }
 
     if (_glfw.x11.disabledCursorWindow)
-        centerCursor(_glfw.x11.disabledCursorWindow);
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/x11_window.c:
```diff
@@ -2234,7 +2236,10 @@ void _glfwPlatformSetCursorMode(_GLFWwindow* window, int mode)
+        if(_glfw.hints.window.centerCursor)
+        {
+                centerCursor(window);
+        }
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

elmindreda commented on src/win32_window.c:
```diff
@@ -1506,7 +1509,10 @@ void _glfwPlatformSetCursorMode(_GLFWwindow* window, int mode)
+        if (_glfw.hints.window.centerCursor)
+        {
+                centerCursor(window);
+        }
```
This relates to `GLFW_CURSOR_DISABLED` and should not be included.

liam-middlebrook commented:
@elmindreda forgot to ping back, do the changes look good now?
